### PR TITLE
perf(qwen3.8): add FA4 QSA backend for H100

### PR DESCRIPTION
**Pull request by HuiyingLi**
# What does this PR do ?

Add an opt-in FlashAttention 4 backend for Qwen3.8 FlashNext's route-sparse QSA on H100. The adapter passes the existing selected-token membership to FA4's public `flash_attn_func`; FA4 computes attention forward, backward, and autograd.

Enable it with the model-owned backend configuration:

```yaml
model:
  backend:
    _target_: nemo_automodel.components.models.qwen3_8_flash_next.backend.Qwen3_8_FlashNextBackendConfig
    attn: cute
```

# Changelog

- Add a model-local FA4 adapter with lazy optional imports and an explicit SM90 / BF16 / head-dimension-256 contract.
- Use PyTorch to build byte membership and forward/backward block lists. Only discrete preprocessing uses `torch.compile(dynamic=False)`. A small CuTe `mask_mod` callback reads membership inside FA4.
- Preserve route-set semantics for duplicates, invalid IDs, empty rows, packed document boundaries, physical padding gaps, and local-query/global-KV coordinates.
- Add `Qwen3_8_FlashNextBackendConfig` and mapping/YAML resolution through the existing model resolver hook. Keep the shared `BackendConfig` unchanged.
- Add CPU contract/configuration tests and H100 output, gradient, activation-checkpoint, stream, and QSA-layer parity coverage.

# Before your PR is "Ready for review"

**Pre checks**:

- [x] Read and followed contributor guidelines.
- [x] Added necessary tests.
- [x] Included usage and dependency requirements in this PR description.

Validation on DFW at head `7b816a33eb3ec19bf6a356e4bd2265b2e56da592`:

- Qwen3.8 CPU model suite: **99 passed, 2 skipped**, with CUDA hidden and `TORCH_COMPILE_DISABLE=1` for CPU math tests.
- H100 functional suite: **11 passed**, with compilation enabled. Checks attention output and all Q/K/V gradients against an independent dense FP64 oracle, plus QSA-layer parameter gradients against the PyTorch implementation.
- GPU coverage includes int32/int64 routes, large invalid IDs, duplicates, empty rows/tiles, full tiles, strided inputs, packed-document gaps, non-reentrant checkpointing, a non-default stream, local-query gradient composition, and 4K route shapes.
- Ruff format/lint and `git diff --check` passed; merge check against current main was clean.
- The focused missing-dependency error test also passed after removing its documentation link.

Tested with FA4 commit `ce088ab9ce0fc0434dcd8afa0a791da9fcc3a820` (the existing Dockerfile pin), CUTLASS DSL 4.6.2, TVM FFI 0.1.11, and PyTorch `2.13.0a0+8145d630e8.nv26.06`.

# Additional Information

The initial backend supports SM90 BF16 D256, first-order autograd, and non-reentrant checkpointing. Deterministic-algorithm mode is rejected because backward uses atomic reductions. Membership storage scales with batch × queries × padded keys. Blackwell and outer layer/model compilation are not enabled by this change.

Existing context-parallel collectives are unchanged. Local-query/global-KV arithmetic is tested on H100; this isolated branch has no new multi-node packed-CP or EP64 MFU result. The earlier combined-experiment MFU numbers should not be attributed to this PR.

Related: #3923 adds the independent DeepSelect indexer backend. This PR changes the attention implementation. Both introduce the model-owned `backend.py`, so integrating both requires combining its attention choice and selector field.

Validation artifacts: `/lustre/fsw/portfolios/coreai/users/huiyingl/qwen38next/fa4_qsa_pr_20260916/` on DFW.

### End-to-end training throughput

Benchmark configuration: [examples/llm_benchmark/qwen/qwen3_8_flash_next_fa4_ep64.yaml](https://github.com/NVIDIA-NeMo/Automodel/blob/4efdad838b1ed32f8367f0ebb1deb9384fb49e24/examples/llm_benchmark/qwen/qwen3_8_flash_next_fa4_ep64.yaml). Uses the standard training loop with 20 warmup steps and 30 measurement steps. W&B logging is opt-in.

Config validation on DFW passed: example YAML lint, typed config construction, all 55 recorded model-config fields, and mock batches of shape `[1, 8192]`. This config-only addition was not rerun on EP64.

Qwen3.8 FlashNext full-model training (48 layers), 8 nodes × 8 H100, EP64, sequence length 8192, LBS 1 / GBS 64, BF16, mock data with `fake_balanced_gate=true`. Ablation based on main `9d875a6d3`; variants ran in the same allocation, with 30 measured steps after 20 warmup steps.

| Variant | TPS/GPU | Throughput gain |
|---|---:|---:|
| [Baseline: Flex QSA](https://wandb.ai/Nemo-automodel/bdelta/runs/1k0ho1x2) | 1,844.56 | — |
| [FA4 QSA](https://wandb.ai/Nemo-automodel/bdelta/runs/rfndujyr) | **2,083.81** | **+12.97%** |

Both variants use the Torch selector and the same indexer scheduling (`query_chunk_size=128`).

TPS/GPU counts training label tokens over the complete training-loop time, divided by 64 GPUs.

### CODE REVIEW

**Comment by HuiyingLi** [comment]
### End-to-end training throughput

Qwen3.8 FlashNext full-model training (48 layers), 8 nodes × 8 H100, EP64, sequence length 8192, LBS 1 / GBS 64, BF16, mock data with `fake_balanced_gate=true`. Ablation based on main `9d875a6d3`; variants ran in the same allocation, with 30 measured steps after 20 warmup steps.

| Variant | TPS/GPU | Throughput gain |
|---|---:|---:|
| [Baseline: Flex QSA](https://wandb.ai/Nemo-automodel/bdelta/runs/1k0ho1x2) | 1,844.56 | — |
| [FA4 QSA](https://wandb.ai/Nemo-automodel/bdelta/runs/rfndujyr) | **2,083.81** | **+12.97%** |

Both variants use the Torch selector and the same indexer scheduling (`query_chunk_size=128`).

TPS/GPU counts training label tokens over the complete training-loop time, divided by 64 GPUs.
